tsne/bh_sne_src/tsne.cpp

```diff
@@ -758,41 +806,92 @@ void run(double* X, int N, int D, double* Y, double perplexity, double theta,
       if (iter == 0)
         fprintf(stderr, "Iteration %d: error is %f\n", iter + 1, C);
       else {
-        total_time += (float)(end - start) / CLOCKS_PER_SEC;
+        elapsed = (float)(end - start) / CLOCKS_PER_SEC;
+        total_time += elapsed;
+        residual_time += elapsed;
         fprintf(stderr,
                 "Iteration %d: error is %f (50 iterations in %4.2f seconds)\n",
-                iter, C, (float)(end - start) / CLOCKS_PER_SEC);
+                iter, C, residual_time);
+        residual_time = 0;
       }
       start = clock();
     }
   }
   end = clock();
-  total_time += (float)(end - start) / CLOCKS_PER_SEC;
+  elapsed = (float)(end - start) / CLOCKS_PER_SEC;
+  total_time += elapsed;
+  residual_time += elapsed;
+
+  if (iter >= max_iter) {
+    fprintf(stderr, "Fitting performed in %4.2f seconds.\n", total_time);
+    return true;
+  }
+  return false;
+}
+
+unique_ptr<TSNEState> make_tsne(double* X, int N, int D, double* Y, int no_dims,
+                                double perplexity, double theta, int rand_seed,
+                                bool skip_random_init, double* init,
+                                bool use_init, int max_iter,
+                                int stop_lying_iter, int mom_switch_iter) {
+  switch (no_dims) {
+    case 2:
+      return make_tsne_impl<2>(X, N, D, Y, perplexity, theta, rand_seed,
+                               skip_random_init, init, use_init, max_iter,
+                               stop_lying_iter, mom_switch_iter);
+    case 3:
+      return make_tsne_impl<3>(X, N, D, Y, perplexity, theta, rand_seed,
+                               skip_random_init, init, use_init, max_iter,
+                               stop_lying_iter, mom_switch_iter);
+    default:
+      throw "invalid dimension";
+  }
+}
 
-  fprintf(stderr, "Fitting performed in %4.2f seconds.\n", total_time);
+bool TSNEState::step_by(int step) {
+  switch (no_dims) {
+    case 2:
+      return step_by_impl<2>(step);
+    case 3:
+      return step_by_impl<3>(step);
+    default:
+      throw "invalid dimension";
+  }
 }
 
 }  // namespace
 
 extern "C" {
+struct TSNE* DLL_PUBLIC init_tsne(double* X, int N, int D, double* Y,
+                                  int no_dims, double perplexity, double theta,
+                                  int rand_seed, bool skip_random_init,
+                                  double* init, bool use_init, int max_iter,
+                                  int stop_lying_iter, int mom_switch_iter) {
+  assert(no_dims == 2 || no_dims == 3);
+  auto tsne = make_tsne(X, N, D, Y, no_dims, perplexity, theta, rand_seed,
+                        skip_random_init, init, use_init, max_iter,
+                        stop_lying_iter, mom_switch_iter);
+  return (TSNE*)tsne.release();
```

Comment:
I think the newer cast syntax is more clear here in terms of order of operations.
```suggestion
  return static_cast<TSNE*>(tsne.release());
```
At least I _think_ `static_cast` will work there?  Possibly might need `reinterpret_cast`?  Better to use `static_cast` where allowed.

Reply:
I will figure it out and use the right one.

Reply:
Have to use `reinterpret_cast` for pointer to pointer casts (except in case of child -> parent class).